miner/remote_proposals.go

```diff
@@ -0,0 +1,219 @@
+	var eg errgroup.Group
+	prepareDisabled := pb.cfg.activeSet.Tries == 0 || pb.cfg.activeSet.RetryInterval == 0
+	if prepareDisabled {
+		pb.logger.Warn("activeset will not be prepared in advance")
+	}
+	for {
+		select {
+		case <-ctx.Done():
+			eg.Wait()
+			return nil
+		case <-pb.clock.AwaitLayer(next):
+			current := pb.clock.CurrentLayer()
+			if current.Before(next) {
+				pb.logger.Info("time sync detected, realigning ProposalBuilder",
+					zap.Uint32("current", current.Uint32()),
+					zap.Uint32("next", next.Uint32()),
+				)
+				continue
+			}
+			next = current.Add(1)
+			ctx := log.WithNewSessionID(ctx)
+
+			if current <= types.GetEffectiveGenesis() {
+				continue
+			}
+			if err := pb.build(ctx, current); err != nil {
+				pb.logger.Warn("failed to build proposal",
+					log.ZContext(ctx),
+					zap.Uint32("lid", current.Uint32()),
+					zap.Error(err),
+				)
+			}
+
+			pb.clean(current)
+		}
+	}
+}
+
+func (pb *RemoteProposalBuilder) clean(layer types.LayerID) {
+	var vals []types.EpochID
+	lim := layer.GetEpoch() - 2
+	for k := range pb.epochEligibilities {
+		if k <= lim {
+			vals = append(vals, k)
+		}
+	}
+	for _, v := range vals {
+		delete(pb.epochEligibilities, v)
+	}
+}
+
+func (pb *RemoteProposalBuilder) build(ctx context.Context, layer types.LayerID) error {
+	pb.signers.mu.Lock()
+	signers := maps.Values(pb.signers.signers)
+	pb.signers.mu.Unlock()
+	for _, signer := range signers {
+		proposal, nonce, err := pb.nodeSvc.Proposal(ctx, layer, signer.signer.NodeID())
+		if err != nil {
+			pb.logger.Error("get partial proposal", zap.Error(err))
+			continue
+		}
+		if proposal == nil {
+			// this node signer isn't eligible this epoch, continue
+			pb.logger.Info("node not eligible on this layer. will try later")
+			continue
+		}
+		bcn, err := pb.nodeSvc.Beacon(ctx, layer.GetEpoch())
+		if err != nil {
+			pb.logger.Error("get beacon", zap.Error(err))
+			continue
+		}
+		var (
+			elig uint32
+			ok   bool
+		)
+		if proposal.Ballot.EpochData != nil {
+			elig, ok = pb.epochEligibilities[layer.GetEpoch()]
+			if !ok {
+				pb.epochEligibilities[layer.GetEpoch()] = proposal.Ballot.EpochData.EligibilityCount
+				elig = proposal.Ballot.EpochData.EligibilityCount
+			}
+		} else {
+			elig, ok = pb.epochEligibilities[layer.GetEpoch()]
+			if !ok {
+				panic("missing epoch eligibilities")
+			}
+		}
+
+		proofs := calcEligibilityProofs(
+			signer.signer.VRFSigner(),
+			layer.GetEpoch(),
+			bcn,
+			types.VRFPostIndex(nonce),
+			elig,
+			pb.cfg.layersPerEpoch,
+		)
+
+		eligibilities, ok := proofs[layer]
+		if !ok {
+			// not eligible in this layer, continue
+			pb.logger.Info("node not eligible in this layer, will try later")
+			continue
+		}
+
+		rng := rand.New(rand.NewSource(time.Now().UnixNano()))
+		rng.Shuffle(len(proposal.TxIDs), func(i, j int) {
+			proposal.TxIDs[i], proposal.TxIDs[j] = proposal.TxIDs[j], proposal.TxIDs[i]
+		})
```

Comment:
The existing proposal shuffling in the `ConservativeState` is more sophisticated and I think this must be consistent, but I'm not 100% sure. @fasmat do you know?

Reply:
It is, but it also relies on shuffling the entire mempool (it pops out the entire mempool every time). I converged on just using a simple shuffle as some sort of a convergence point.

Reply:
We shuffle Tx in multiple places. If I remember correctly the place where order matters is in `getBlockTXs` in `blocks/utils.go`. The order of transactions in a proposal does not matter, besides ideally it should be random for all proposals even if created from the same node but for different identities - which currently is not the case when the node runs on windows (due to the low resolution of `time.Now()`).

Reply:
However the nonce order per principal is still important for the order in proposals, so after shuffling them the TXs are swapped to ensure that lower nonces come first:

After shuffling txs we get this as an example

```
principal1, nonce 5
principal2, nonce 3
principal2, nonce 2
principal1, nonce 3
principal1, nonce 4
```

and then txs are swapped to ensure nonce order is preserved like this:

```
principal1, nonce 3
principal2, nonce 2
principal2, nonce 3
principal1, nonce 4
principal1, nonce 5
```

Reply:
Interestingly we don't have a test that this function (`ShuffleWithNonceOrder`) actually does this correctly 😅

Reply:
Er... I looked at it another time and indeed this seems like it could be a problem if we don't use the existing implementation. I think that maybe PredictBlock should be removed and we could just use the existing implementation. The concern was that we need to yield different block proposals to different requesters, that's why the crude selection of transaction. However, in this case this would potentially be erroneous because we do need to take into account the nonce value sorting that is done in the shuffle function.

Reply:
Probably just having the random number generator as-is coming from the timestamp should be good enough to produce different results to different requesters.

Reply:
@acud Instead of the timestamp maybe you can use the `NodeID` in some way as a seed? We want to guarantee that proposals from different identities have differently shuffled transactions (just nonce order preserved). If we use the timestamp there is a chance that multiple identities that are eligible at the layer generate identical proposals (especially on windows where `time.Now()` has a terrible resolution).

Reply:
So the shuffling occurs on the node side, not the smesher service side. I'm not sure what are the expectations on where the _nodes_ will be run and on which platform. If we expect the smeshers to be windows - no problem. If we expect the nodes to be windows - I can add an extra entropy byte slice that could be passed to the random number generator and we could add the node ID there in that way. Honestly unless a majority of the _nodes_ run on windows in this case, I don't think this should be a problem... and even then, it would only be exacerbated when there's a lot of requests for proposals landing exactly on the same time.